Minor bug fixes for v0.6

borisliao opened the pull request:
Fixes most active bugs in the issues. Look at individual commits for issue numbers.

borisliao committed:
Fixed sharing of playlists #22

borisliao committed:
Renamed list to playlist

borisliao committed:
Merge branch 'master' of https://github.com/borisliao/mbox into v0.6

borisliao committed:
Fix #6: bot will update paused footer state

borisliao committed:
Fix #7: Play next song if in a channel & not pause

borisliao committed:
Fixed #21

borisliao committed:
Fix #19 by messaging the owner and leaving

borisliao committed:
Fix #14 by checking current song exits

borisliao committed:
#21 Delete message if its from another bot.

borisliao committed:
Added changes to changelog

AvJain commented on src/music/element/MusicQueue.py:
```diff
@@ -9,8 +9,11 @@
     """
-    def __init__(self, active_embed: ChatEmbed, client: discord.Client, list: List[MusicSource] = []) -> None:
-        self.list = list
+    def __init__(self, active_embed: ChatEmbed, client: discord.Client, playlist: List[MusicSource] = None) -> None:
```
Any reason why you had to set playlist to None instead in the arguments instead of just having it set as an empty list by default

Karan-Jain commented on src/music/element/MusicQueue.py:
```diff
@@ -9,8 +9,11 @@
-    def __init__(self, active_embed: ChatEmbed, client: discord.Client, list: List[MusicSource] = []) -> None:
-        self.list = list
+    def __init__(self, active_embed: ChatEmbed, client: discord.Client, playlist: List[MusicSource] = None) -> None:
+        if playlist == None:
```
Not a biggie, but to make the code cleaner you can probably do this.
```suggestion
  self.playlist =[]
  if playlist:
     self.playlist = playlist
```

AvJain commented on src/music/element/MusicQueue.py:
```diff
@@ -9,8 +9,11 @@
     """
-    def __init__(self, active_embed: ChatEmbed, client: discord.Client, list: List[MusicSource] = []) -> None:
-        self.list = list
+    def __init__(self, active_embed: ChatEmbed, client: discord.Client, playlist: List[MusicSource] = None) -> None:
```
Any reason why you had to set playlist to None in the arguments instead of just having it set as an empty list by default

borisliao replied:
That was the problem that I fixed when servers were sharing queue's (list variable) with each other since python is strictly pass by reference only. When I initialized a queue, the default behavior was to use the same empty list address for all the queues for all the servers. From what I researched default variables in classes are counted as class variables, and anything inside the function is counted as instance variables.

AvJain replied:
gotcha, that makes sense

AvJain commented on src/music/player.py:
```diff
@@ -297,10 +298,11 @@ async def play_youtube(self, link: str):
-                        self.paused = False
 
+                        # If the player is not playing because it just came in to the channel (not because of being paused), advance the track head to the next (just added) song
                         if not self.connected_client.is_playing():
```
can this not be done using an and statement?

borisliao replied:
Do you mean like this? It's not a problem, whichever is the most clear.
```suggestion
                        if not self.connected_client.is_playing() and not self.connected_client.is_paused():
```

AvJain replied:
yup I prefer it be more concise, the less nested something is the better in my opinion

Karan-Jain commented on src/music/player.py:
```diff
@@ -314,9 +316,10 @@ def on_read(ms, non_music):
-                                    self.add_to_footer(source= 'Youtube 📁', icon_url=YOUTUBE_ICON)
-                                    asyncio.run_coroutine_threadsafe(self.messenger.gui['player'].update(), self.connected_client.loop)
+                                if self.playlist.current():
+                                    if(self.playlist.current().info == info):
```
Remove the  parenthesis around this if (Best practice to generally not have them)

AvJain replied:
> Remove the parenthesis around this if
I would say to stay consistent to either use parenthesis for all if's or not use them at all

borisliao replied:
```suggestion
                                    if self.playlist.current().info == info:
```
I agree. No parenthesis for just a single mundane check.

AvJain commented on main.py:
```diff
@@ -125,13 +125,18 @@ async def on_typing(channel, user, when):
+        await src.preinitialization.generate_profile(guild, mbox, profiles)
+        for profile in profiles:
+            await profile.setup()
+    except discord.errors.Forbidden:
```
Are 403's the only error we see due to not having permissions? How about 401's?

Karan-Jain replied:
Yea I think 401 is also something we definitely want to have

borisliao replied:
It looks like 403's are the only one that it gives when there is a permission error.
https://discord.com/developers/docs/topics/opcodes-and-status-codes#http
This is why I think the discord python library gives this 403 its own custom error instead of a generic discord.HTTPException
https://discordpy.readthedocs.io/en/stable/api.html?highlight=forbidden#discord.Forbidden

AvJain replied:
ok got it. Seems like it doesn't define logic related to 401's then. Thanks for clarifying

Karan-Jain commented on src/constants.py:
```diff
@@ -9,6 +9,7 @@
 USAGE_TEXT = 'Send a search phrase or a link to play a song!'
```
Why not use Enums instead of constants?

borisliao replied:
My understanding of enums is its for variables that you want to enumerate. 
Unless you are talking about put all the program constants in one enum. So instead of making the files import * from constants.py, do you mean you can import one enum variable from constants.py?

Karan-Jain replied:
Yup, that's exactly what I am talking about!

borisliao replied:
I never really thought of enums like that but it definitely would be better then importing * from a file. Probably will put this as a refactor issue instead of fixing it in the minor bug fixes PR.

borisliao replied:
Issue created #46

AvJain reviewed (commented):
I like that you added the invite link to the constants file. Just had a few questions that I needed clarification on from my side. Other than that everything else looks good

Karan-Jain reviewed (commented):
I gave some suggestions on  couple of minor things that can tweaked

borisliao committed:
Update src/music/element/MusicQueue.py

Co-authored-by: Karan Jain <anukaran.jain@gmail.com>

borisliao committed:
Update src/music/player.py

Exsects commented on main.py:
```diff
@@ -125,13 +125,18 @@ async def on_typing(channel, user, when):
 @mbox.event
-async def on_guild_join(guild):
+async def on_guild_join(guild: discord.Guild):
     logging.info('Joined Server: {0.name}'.format(guild))
```
```suggestion
    logging.info(f'Joined Server: {guild.name}')
```
As of Python 3.6, a better alternative to using String format() is to use f-strings.

https://stackoverflow.com/questions/42379271/format-function-alternative-in-python

Exsects commented on main.py:
```diff
@@ -125,13 +125,18 @@ async def on_typing(channel, user, when):
+    except discord.errors.Forbidden:
+        owner_member: discord.Member = guild.owner
+        if owner_member:
+            await owner_member.send(content="You or a member in server ["+ guild.name + "] added me without the right permissions. Try adding me again with the link: " + INVITE_LINK_FORMAT.format(mbox.user.id))
```
```suggestion
content= f"You or a member in server  {guild.name} added me without the right permissions. Try adding me again with the link: " + INVITE_LINK_FORMAT.format(mbox.user.id)
await owner_member.send(content=content)
```
Similar to my earlier suggestion, I would recommend using f-strings. Also, I would suggest you to isolate the string by assigning it to its own variable. Just for my understanding, where is INVITE_LINK.format() coming from?

borisliao replied:
Yeah, f-strings are defiantly more concise.

> Just for my understanding, where is INVITE_LINK.format() coming from?
It comes from the new constant variable in constants.py. This is planned to be changed per #46.

Exsects commented on src/music/player.py:
```diff
@@ -314,9 +316,10 @@ def on_read(ms, non_music):
-                                    self.add_to_footer(source= 'Youtube 📁', icon_url=YOUTUBE_ICON)
-                                    asyncio.run_coroutine_threadsafe(self.messenger.gui['player'].update(), self.connected_client.loop)
+                                if self.playlist.current():
+                                    if self.playlist.current().info == info:
```
```suggestion
                                if self.playlist.current() and self.playlist.current().info == info:
```
Think this would be better instead of nested if statements.

borisliao commented on src/music/player.py:
```diff
@@ -297,10 +298,11 @@ async def play_youtube(self, link: str):
+                        # If the player is not playing because it just came in to the channel (not because of being paused), advance the track head to the next (just added) song
                         if not self.connected_client.is_playing():
-                            self.next()
+                            if not self.connected_client.is_paused():
+                                self.next()
```
```suggestion
                            self.next()
```

borisliao committed:
Remove nested if statements from Code Reivew

https://github.com/borisliao/mbox/pull/43/files/0f5dfb004048016050305822c759244638f9c2c9#r627903538

borisliao committed:
Update src/music/player.py

borisliao committed:
Update main.py

Co-authored-by: Eric Wong <38166320+Exsects@users.noreply.github.com>

borisliao committed:
Update main.py

Co-authored-by: Eric Wong <38166320+Exsects@users.noreply.github.com>

borisliao committed:
Update src/music/player.py removing nested if statement

Co-authored-by: Eric Wong <38166320+Exsects@users.noreply.github.com>

borisliao committed:
Fix code review indenting issues

AvJain reviewed (approved):
Changes look good to me

Karan-Jain reviewed (approved):
All good!
